### PR TITLE
[DTensor] Support user-supplied Generator for random ops

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159933

If the user provides a generator kwarg to a random op (e.g.
nn.init.uniform_(..., generator=my_generator)), we can still advance
that generator's state in a SPMD-global way so that each local-tensor
gets appropriate values and the generator advances to the same state as
if it had operated on the full tensor.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @pragupta

### CODE REVIEW

File: test/distributed/tensor/test_random_ops.py [hunk 1]
```diff
@@ -87,6 +87,21 @@ def test_init_ops(self):
             self._run_init_op(torch.randn_like, dtype=dtype)
             self._run_init_op(torch.randint_like, low=0, high=100, dtype=dtype)
 
+    @with_comms
+    def test_init_with_user_generator(self):
```
Comment (wconstab) [inline test/distributed/tensor/test_random_ops.py:92]: Missing gpu annotation (ci fail)

File: torch/distributed/tensor/_random.py [hunk 2]
```diff
@@ -146,7 +146,9 @@ def set_seed(self, name: str, seed: int) -> None:
         )
         self.rng_states[name] = torch.cat([seed_tensor, offset_tensor])
 
-    def _distribute_region(self, spec: DTensorSpec):
+    def _distribute_region(
+        self, spec: DTensorSpec, generator: Optional[torch.Generator]
```
Comment (wconstab) [inline torch/distributed/tensor/_random.py]: Missing =None

File: torch/distributed/tensor/_dispatch.py [hunk 3]
```diff
@@ -197,8 +197,19 @@ def dispatch(
...
                 )
+
+                # If the user provided a generator, we hook it up to our RNG manager, but we also pop it from kwargs
+                # so the op_call does not directly use it (we want op_call to fall back to the 'default' which is
+                # our RNG manager)
+                maybe_user_generator = op_info.local_kwargs.pop("generator", None)
+                assert maybe_user_generator is None or isinstance(
+                    maybe_user_generator, torch.Generator
+                )
+                # maybe_user_generator = None
```
Comment (XilunWu) [inline torch/distributed/tensor/_dispatch.py:207]: remove comment

File: torch/distributed/tensor/_random.py [hunk 4]
```diff
@@ -191,7 +193,17 @@ def _manual_seed(self, parallel_seed: int) -> None:
...
+        self, spec: DTensorSpec, generator: Optional[torch.Generator] = None
+    ):
+        g_name = "parallel-rng"
+        if generator is not None:
+            # This is a little hacky, but for any user-passed generator, we store its state under a unique key,
+            # not because we need to keep a copy of it but because its the easiest way to make it work with the
+            # existing set/get APIs
+            g_name = str(id(generator))
+            if g_name not in self.rng_states:
+                self.rng_states[g_name] = generator.get_state()
```
Comment (XilunWu) [inline torch/distributed/tensor/_random.py]: this is a slight behavior divergence over using the default RNG or the user-specified RNG. Either we require users to call our `manual_seed()` API with the RNG passed in, or we optimistically assume users know what they're doing and are responsible for initializing the RNG with the right seed value across ranks (which is what we're doing here).

**Review by XilunWu — COMMENTED** [review]
Overall LGTM. Let me know if you need quick unblock.

We need to explicitly tell users to initialize the generator passed in with the same seed,  or we add this functionality to our `manual_seed` API but users still need to be aware of this.

File: torch/distributed/tensor/_random.py (hunk 4 above)
Comment (wconstab) [reply torch/distributed/tensor/_random.py]: I think I like the behavior this way- I don't want to introduce a collective on every op where a user supplies an RNG. 

I will add documentation to dtensor docs stating that it is the user's responsibility to ensure the passed generator has the same state on every spmd rank.

File: torch/distributed/tensor/_random.py [hunk 5]
```diff
@@ -191,7 +193,17 @@ def _manual_seed(self, parallel_seed: int) -> None:
...
+    def _distribute_region(
+        self, spec: DTensorSpec, generator: Optional[torch.Generator] = None
+    ):
+        g_name = "parallel-rng"
+        if generator is not None:
+            # This is a little hacky, but for any user-passed generator, we store its state under a unique key,
+            # not because we need to keep a copy of it but because its the easiest way to make it work with the
+            # existing set/get APIs
+            g_name = str(id(generator))
+            if g_name not in self.rng_states:
```
Comment (wconstab) [inline torch/distributed/tensor/_random.py]: I will also add 2 more test cases and fix one bug here:

If the user changed the seed of their generator after using it with dtensor, we'd cache the first seed and use that forever. To prevent this I'm going to pop the temporary generator back out of self.rng_stated at the end of this context, so we add it fresh again every time.

This makes me want to do a refactor, I think we should not be storing states like this. We should probably just store a ref to our own generator and then use it.

File: test/distributed/tensor/test_random_ops.py [hunk 6]
```diff
@@ -87,6 +87,39 @@ def test_init_ops(self):
...
+        torch.manual_seed(42)
+        rng = torch.Generator(device="cuda").manual_seed(42)
+        t1 = torch.distributed.tensor.empty(
+            (2, 3), device_mesh=device_mesh, placements=[Shard(0)]
+        )
+        t2 = torch.distributed.tensor.empty(
+            (2, 3), device_mesh=device_mesh, placements=[Shard(0)]
+        )
+        for i in range(2):
+            print(f"{i=}")
```
Comment (XilunWu) [inline test/distributed/tensor/test_random_ops.py:103]: this can be removed since we have 'i' in assert error

**Review by fduwjj — APPROVED** [review]

File: torch/distributed/tensor/_random.py (hunk 4 above)
Comment (lucasb-eyer) [reply torch/distributed/tensor/_random.py]: the original feature request comes from @akolesnikoff and myself. We noticed the issue specifically because we were wondering about and comparing behaviour of passing in same vs different seeded RNGs across ranks. So at least from our perspective, yes, we are intentional.

Comment (XilunWu) [reply torch/distributed/tensor/_random.py]: I see. 

> passing in same vs different seeded RNGs across ranks

I assume you're trying to pass same RNGs for Data Parallel and different RNGs for Model Parallel (let me know if this is not the case).

File: test/distributed/tensor/test_random_ops.py [hunk 7]
```diff
@@ -87,6 +87,38 @@ def test_init_ops(self):
...
+        #     they may be surprised that it has no effect on DTensor.  They must instead call this private API
+        #     (`torch.distributed.tensor._random._rng_tracker._manual_seed`)
+        # (b) If we try to match the semantics of (a) with a user-supplied RNG, they may be very surprised to find that
+        #     their RNG object never advances its state after using it with DTensor.
+        # torch.distributed.tensor._random._rng_tracker._manual_seed(55)
+        # rng.manual_seed(55)
+        # torch.nn.init.uniform_(t1, 0.0, 1.0)
+        # torch.nn.init.uniform_(t2, 0.0, 1.0, rng)
+        # self.assertEqual(t1.full_tensor(), t2.full_tensor())
+
```
Comment (XilunWu) [inline test/distributed/tensor/test_random_ops.py:121]: clean up this comment, maybe move to _random.py?

**Review by XilunWu — APPROVED** [review]
this API change lgtm

File: test/distributed/tensor/test_random_ops.py (hunk 7 above)
Comment (wconstab) [reply test/distributed/tensor/test_random_ops.py:121]: i think its ok to leave it here for now.  i still need to get some agreement on if we're changing the default rng behavior.  then i'd make another PR to do that and i can remove this TODO and enable this part of the test

File: torch/distributed/tensor/_random.py [hunk 8]
```diff
@@ -191,7 +193,17 @@ def _manual_seed(self, parallel_seed: int) -> None:
...
-    def _distribute_region(self, spec: DTensorSpec):
+    def _distribute_region(
+        self, spec: DTensorSpec, generator: Optional[torch.Generator] = None
+    ):
+        g_name = "parallel-rng"
+        if generator is not None:
+            # This is a little hacky, but for any user-passed generator, we store its state under a unique key,
+            # not because we need to keep a copy of it but because its the easiest way to make it work with the
+            # existing set/get APIs. We also ensure we remove it from rng_states after each _distribute_region.
+            g_name = "user-passed-generator"
```
Comment (wanchaol) [inline torch/distributed/tensor/_random.py:204]: I wonder if it could just be str(generator) as the key?

**Review by wanchaol — APPROVED** [review]
lgtm

File: torch/distributed/tensor/_random.py (hunk 8 above)
Comment (wconstab) [reply torch/distributed/tensor/_random.py:204]: i can try this. i think i'll land as-is and try experimenting more with this in the next PR.

I already tried using `id(generator)` as the key- this did not work, and I suspect we are making copies of the python wrapper at some point in our bindings or our dispatching layer, leading to the ID changing.  I do notice that the `str(generator)` prints what looks like a memory address for the underlying CPP object, so it might indeed be more stable and fix my issue.  Thanks for the suggestion!
